File: frontend/src/api/gates.ts
import request from '@/utils/request'

export interface Gate {
  id: string
  name: string
  description?: string
  pipeline_id: string
  pipeline_name?: string
  type: string
  conditions: GateCondition[]
  is_active: boolean
  created_by?: string
  created_at: string
  updated_at: string
}

export interface GateCondition {
  metric: string
  operator: string
  value: string
  unit?: string
}

export interface GateCreate {
  name: string
  description?: string
  pipeline_id: string
  type: string
  conditions: GateCondition[]
}

export interface GateUpdate {
  name?: string
  description?: string
  pipeline_id?: string
  type?: string
  conditions?: GateCondition[]
  is_active?: boolean
}

export interface GateCheckResult {
  gate_id: string
  gate_name: string
  passed: boolean
  conditions: {
    metric: string
    expected: string
    actual: string
    passed: boolean
  }[]
  checked_at: string
}

export const gatesApi = {
  // Get all gates
  getGates() {
    return request.get<Gate[]>('/v1/gates/')
  },

  // Get gate by ID
  getGate(gateId: string) {
    return request.get<Gate>(`/v1/gates/${gateId}`)
  },

  // Create gate
  createGate(data: GateCreate) {
    return request.post<Gate>('/v1/gates/', data)
  },

  // Update gate
  updateGate(gateId: string, data: GateUpdate) {
    return request.put<Gate>(`/v1/gates/${gateId}`, data)
  },

  // Delete gate
  deleteGate(gateId: string) {
    return request.delete<{ message: string }>(`/v1/gates/${gateId}`)
  },

  // Check gate
  checkGate(gateId: string, pipelineRunId: string) {
    return request.post<GateCheckResult>(`/v1/gates/${gateId}/check`, { pipeline_run_id: pipelineRunId })
  },

  // Get gate check history
  getGateChecks(gateId: string) {
    return request.get<GateCheckResult[]>(`/v1/gates/${gateId}/checks`)
  },
}

File: backend/app/api/v1/endpoints/gates.py
from typing import List
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from app.core.database import get_db
from app.core.deps import get_current_active_user
from app.models.user import User

router = APIRouter()


@router.get("/")
async def get_gates(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    # TODO: 实现门禁查询
    return []


@router.get("/{gate_id}")
async def get_gate(
    gate_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    # TODO: 实现单个门禁查询
    return {}


@router.post("/")
async def create_gate(
    gate_data: dict,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    # TODO: 实现门禁创建
    return gate_data


@router.put("/{gate_id}")
async def update_gate(
    gate_id: str,
    gate_data: dict,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    # TODO: 实现门禁更新
    return gate_data


@router.delete("/{gate_id}")
async def delete_gate(
    gate_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    # TODO: 实现门禁删除
    return {"message": "Gate deleted"}


@router.post("/{gate_id}/check")
async def check_gate(
    gate_id: str,
    data: dict,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    # TODO: 实现门禁检查
    return {
        "gate_id": gate_id,
        "gate_name": "Quality Gate",
        "passed": True,
        "conditions": [],
        "checked_at": "2024-01-01T00:00:00Z"
    }


@router.get("/{gate_id}/checks")
async def get_gate_checks(
    gate_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    # TODO: 实现检查历史查询
    return []
